File: apps/mobile/modules/cindy-html-preview/ios/CindyHtmlPreviewModule.swift
import ExpoModulesCore

public class CindyHtmlPreviewModule: Module {
  private var servers: [String: HtmlSnapshotServer] = [:]
  private let queue = DispatchQueue(label: "cindy.html-preview")
  private var foreground = true

  public func definition() -> ModuleDefinition {
    Name("CindyHtmlPreview")
    Events("resourceRequest", "resourceClosed")
    AsyncFunction("start") { (root: String, entry: String, token: String, csp: String, files: [[String]], promise: Promise) in
      do {
        guard self.foreground, self.servers.count < 4, self.servers[token] == nil,
          let cache = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first,
          URL(fileURLWithPath: root).resolvingSymlinksInPath().path.hasPrefix(cache.resolvingSymlinksInPath().path + "/")
        else { throw HtmlSnapshotServer.Failure.invalid }
        let server = try HtmlSnapshotServer(root: root, entry: entry, token: token, csp: csp, files: files, queue: self.queue)
        self.servers[token] = server
        server.start { result in
          switch result {
          case .success(let url): promise.resolve(url)
          case .failure:
            self.servers.removeValue(forKey: token)?.stop()
            promise.reject("HTML_PREVIEW_START_FAILED", "Unable to start HTML preview")
          }
        }
      } catch { promise.reject("HTML_PREVIEW_START_FAILED", "Unable to start HTML preview") }
    }.runOnQueue(queue)
    AsyncFunction("startOnDemand") { (root: String, entry: String, token: String, csp: String, promise: Promise) in
      do {
        guard self.foreground, self.servers.count < 4, self.servers[token] == nil,
          let cache = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first,
          URL(fileURLWithPath: root).resolvingSymlinksInPath().path.hasPrefix(cache.resolvingSymlinksInPath().path + "/")
        else { throw HtmlSnapshotServer.Failure.invalid }
        let server = try HtmlSnapshotServer(root: root, entry: entry, token: token, csp: csp, files: [], queue: self.queue,
          onRequest: { [weak self] id, path in self?.sendEvent("resourceRequest", ["token": token, "id": id, "path": path]) },
          onClose: { [weak self] id in self?.sendEvent("resourceClosed", ["token": token, "id": id]) })
        self.servers[token] = server
        server.start { result in
          switch result {
          case .success(let url): promise.resolve(url)
          case .failure:
            self.servers.removeValue(forKey: token)?.stop()
            promise.reject("HTML_PREVIEW_START_FAILED", "Unable to start HTML preview")
          }
        }
      } catch { promise.reject("HTML_PREVIEW_START_FAILED", "Unable to start HTML preview") }
    }.runOnQueue(queue)
    AsyncFunction("resolveRequest") { (token: String, id: String, filename: String, mime: String, status: Int) -> Bool in
      self.servers[token]?.resolve(id, filename: filename, mime: mime, status: status) ?? false
    }.runOnQueue(queue)
    AsyncFunction("stop") { (token: String) in self.servers.removeValue(forKey: token)?.stop() }.runOnQueue(queue)
    OnAppEntersBackground { self.queue.async { self.foreground = false; self.stopAll() } }
    OnAppEntersForeground { self.queue.async { self.foreground = true } }
    OnDestroy { self.queue.async { self.stopAll() } }
  }

  private func stopAll() {
    let active = Array(servers.values)
    servers.removeAll()
    active.forEach { $0.stop() }
  }
}

File: apps/mobile/modules/cindy-html-preview/android/src/main/java/expo/modules/cindyhtmlpreview/CindyHtmlPreviewModule.kt
package expo.modules.cindyhtmlpreview

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File

class CindyHtmlPreviewModule : Module() {
  private val servers = mutableMapOf<String, HtmlSnapshotServer>()
  private var foreground = true
  override fun definition() = ModuleDefinition {
    Name("CindyHtmlPreview")
    Events("resourceRequest", "resourceClosed")
    AsyncFunction("start") { root: String, entry: String, token: String, csp: String, files: List<List<String>> ->
      synchronized(servers) {
        val cache = requireNotNull(appContext.reactContext).cacheDir.canonicalPath
        require(File(root).canonicalPath.startsWith(cache + File.separator))
        require(foreground && servers.size < 4 && !servers.containsKey(token))
        val server = HtmlSnapshotServer(root, entry, token, csp, files)
        servers[token] = server
        server.start()
      }
    }
    AsyncFunction("startOnDemand") { root: String, entry: String, token: String, csp: String ->
      synchronized(servers) {
        val cache = requireNotNull(appContext.reactContext).cacheDir.canonicalPath
        require(File(root).canonicalPath.startsWith(cache + File.separator))
        require(foreground && servers.size < 4 && !servers.containsKey(token))
        val server = HtmlSnapshotServer(root, entry, token, csp, emptyList(),
          onRequest = { id, path -> sendEvent("resourceRequest", mapOf("token" to token, "id" to id, "path" to path)) },
          onClose = { id -> sendEvent("resourceClosed", mapOf("token" to token, "id" to id)) })
        servers[token] = server
        server.start()
      }
    }
    AsyncFunction("resolveRequest") { token: String, id: String, filename: String, mime: String, status: Int ->
      synchronized(servers) { servers[token]?.resolve(id, filename, mime, status) ?: false }
    }
    AsyncFunction("stop") { token: String -> synchronized(servers) { servers.remove(token)?.stop(); Unit } }
    OnActivityEntersBackground { synchronized(servers) { foreground = false; stopAll() } }
    OnActivityEntersForeground { synchronized(servers) { foreground = true } }
    OnDestroy { stopAll() }
  }
  private fun stopAll() = synchronized(servers) {
    servers.values.forEach { it.stop() }
    servers.clear()
  }
}
